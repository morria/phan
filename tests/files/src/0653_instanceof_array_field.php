<?php

namespace NS653;

/**
 * @param array{field:object} $array
 */
function check_instanceof($array) {
    if ($array['field'] instanceof \ArrayAccess) {
        $array['field']->someMethod();
        return;
    }
    echo strlen($array['field']);
}

/**
 * @param array{0:\stdClass|\Countable} $array
 */
function check_instanceof_negate($array) {
    if ($array[0] instanceof \Countable) {
        var_export($array[0]->count());
        var_export($array['field']->count());
        return;
    }
    var_export($array[0]->count());
}

const FIVE = 5;

/**
 * @param array{5:\stdClass|\Countable} $array
 */
function check_instanceof_negate_2($array) {
    if ($array[FIVE] instanceof \Countable) {
        var_export($array[FIVE]->count());
        var_export($array[4]->count());
        return;
    }
    var_export($array[FIVE]->count());
}

/**
 * @param \stdClass|int $o
 */
function check_var_negate($o) : string {
    if ($o instanceof \stdClass) {
        return $o;
    }
    return $o;
}
