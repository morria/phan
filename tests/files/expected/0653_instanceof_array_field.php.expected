%s:10 PhanUndeclaredMethod Call to undeclared method \ArrayAccess::someMethod
%s:13 PhanTypeMismatchArgumentInternal Argument 1 (string) is object but \strlen() takes string
%s:22 PhanNonClassMethodCall Call to method count on non-class type null
%s:22 PhanTypeInvalidDimOffset Invalid offset "field" of array type array{0:\Countable}
%s:25 PhanUndeclaredMethod Call to undeclared method \stdClass::count
%s:36 PhanNonClassMethodCall Call to method count on non-class type null
%s:36 PhanTypeInvalidDimOffset Invalid offset 4 of array type array{5:\Countable}
%s:39 PhanUndeclaredMethod Call to undeclared method \stdClass::count
%s:47 PhanTypeMismatchReturn Returning type \stdClass but check_var_negate() is declared to return string
%s:49 PhanTypeMismatchReturn Returning type int but check_var_negate() is declared to return string
